pkg/scheduler/cache/cache.go

```diff
@@ -113,13 +142,25 @@ func (cache *schedulerCache) Snapshot() *Snapshot {
 func (cache *schedulerCache) UpdateNodeNameToInfoMap(nodeNameToInfo map[string]*NodeInfo) error {
 	cache.mu.Lock()
 	defer cache.mu.Unlock()
+
+	imageStateSummaries := make(map[string]*ImageStateSummary)
+
 	for name, info := range cache.nodes {
 		if utilfeature.DefaultFeatureGate.Enabled(features.BalanceAttachedNodeVolumes) && info.TransientInfo != nil {
 			// Transient scheduler info is reset here.
 			info.TransientInfo.resetTransientSchedulerInfo()
 		}
 		if current, ok := nodeNameToInfo[name]; !ok || current.generation != info.generation {
-			nodeNameToInfo[name] = info.Clone()
+			clone := info.Clone()
+			// Add image state information to the clone for all images in the node
+			for image := range info.imageStates {
```

Comment:
I agree any potential overhead added to the main scheduling loop should be reasoned about carefully - I had it in mind :)

1. The part I am concerned (may be unnecessary) is concurrent access. If we do not do a "clone" here for the image, then reading the imageStates needs to acquire the mutex of the scheduler cache. Currently the mutex is acquired here only: 

https://github.com/silveryfu/kubernetes/blob/cad55e9f27771a64895a7e95b77ea7651c546925/pkg/scheduler/core/generic_scheduler.go#L124

2. Here we are not cloning but instead making an ImageStateSummary for each image state (for each image, this generation happens only once with `imageStateSummaries := make(map[string]*ImageStateSummary)`). The reason is that the original imageState struct can have information we do not need and we need aggregated information - aggregation is relatively cheap here (just len()s).

3. I agree approximated results may be good enough (for clarification, this would allow the imageStates data aggregation happen outside the scheduling loop).

The key difference seems to be "lazy aggregation" vs. "active aggregation". Lazy aggregation is to postpone the aggregation until the scheduling loop (which is the most accurate); whereas active aggregation is less accurate (some nodeInfo will not have the most updated NumNodes and NumTotalNodes when they are cloned). Active aggregation should have little overhead on the scheduling loop as all the imageStateSummaries will be computed beforehand.

Reply:
In one of my recent PRs, I changed the lock in cache to `cache.RWMutex` which should improve multi-read case if it concerns you. Anyway the point that priority is only "best effort" really make sense here :D

Reply:
:) That's really nice @resouer ! Though multi-reader does not concern me - it is the writers to the cache who should never block the readers and yet it is not possible w/o cloning.

@resouer @bsalamat Based on the discussion above, I will revise it to doing "active aggregation" - that is, creating imageStateSummaries each time we SetNode() and pass its reference via NodeInfo (and TotalNumNodes through meta).

This will avoid additional cloning in the scheduling cycle (at the cost of some accuracy - depending on how long the nodes refresh themselves - btw, how long that is by default?)

I will do the revision this weekend. Thank you for the comments!

Reply:
Node heartbeat interval is 10s, and will be refactored to on demand in the future. Should be OK.